Anomaly Detection: Credit Card Fraud Detection

Importing the Libereries

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score


In [ ]:
# Loading the dataset
credit=pd.read_csv('creditcard.csv')

In [ ]:
credit.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0     0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1     0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2     1 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3     1 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4     2 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62    0.0  
1  0.125895 -0.008983  0.014724    2.69    0.0  
2 -0.139097 -0.055353 -0.059752  378.66    0.0  
3 -0.221929  0.062723  0.061458  123.50    0.0  
4  0.502292  0.219422  0.215153   69.99    0.0  

[5 rows x 31 columns]

In [ ]:
credit.tail()

Time        V1        V2        V3        V4        V5        V6  \
87235  61595 -0.788289 -0.270922  1.426067 -1.557292  0.326118 -1.139993   
87236  61595  1.510898 -1.237081  0.055205 -1.596874 -1.328718 -0.290028   
87237  61595 -0.903199  0.920795  1.087351  1.206105 -0.123057  0.673090   
87238  61596 -1.136232  0.825319  0.466206 -1.296195  0.493619 -0.109541   
87239  61596 -6.759521  5.572257 -0.951531 -3.194834       NaN       NaN   

             V7        V8        V9  ...       V21       V22       V23  \
87235  0.191611  0.055201  1.245247  ...  0.138402  0.528514 -0.099555   
87236 -1.184932  0.088371 -1.669287  ... -0.238966 -0.474265 -0.059932   
87237  0.369554  0.678393 -0.830880  ...  0.097581  0.188902 -0.061636   
87238  0.571839  0.355682 -0.873281  ...  0.100479  0.027285 -0.329304   
87239       NaN       NaN       NaN  ...       NaN       NaN       NaN   

            V24       V25       V26       V27       V28  Amount  Class  
87235  0.436539  0.305689 -0.779462  0.082087  0.048889    9.95    0.0  
87236 -0.588342  0.378212 -0.165141  0.006521 -0.001020   25.00    0.0  
87237 -0.304928  0.168362 -0.186627 -0.007386  0.033238   80.51    0.0  
87238 -0.868148  0.200414  0.944678 -0.269153 -0.013332   51.34    0.0  
87239       NaN       NaN       NaN       NaN       NaN     NaN    NaN  

[5 rows x 31 columns]

In [ ]:
credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87240 entries, 0 to 87239
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    87240 non-null  int64  
 1   V1      87240 non-null  float64
 2   V2      87240 non-null  float64
 3   V3      87240 non-null  float64
 4   V4      87240 non-null  float64
 5   V5      87239 non-null  float64
 6   V6      87239 non-null  float64
 7   V7      87239 non-null  float64
 8   V8      87239 non-null  float64
 9   V9      87239 non-null  float64
 10  V10     87239 non-null  float64
 11  V11     87239 non-null  float64
 12  V12     87239 non-null  float64
 13  V13     87239 non-null  float64
 14  V14     87239 non-null  float64
 15  V15     87239 non-null  float64
 16  V16     87239 non-null  float64
 17  V17     87239 non-null  float64
 18  V18     87239 non-null  float64
 19  V19     87239 non-null  float64
 20  V20     87239 non-null  float64
 21  V21     87239 non-null  float64
 22

In [ ]:
credit.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        1
V6        1
V7        1
V8        1
V9        1
V10       1
V11       1
V12       1
V13       1
V14       1
V15       1
V16       1
V17       1
V18       1
V19       1
V20       1
V21       1
V22       1
V23       1
V24       1
V25       1
V26       1
V27       1
V28       1
Amount    1
Class     1
dtype: int64

In [ ]:
# Now Handeling the missing values with Imputation method

In [ ]:
# Analysing the distribution of data
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
for col in credit.columns:
    if credit[col].isnull().sum() > 0:
        # Check if the column is numeric before calculating the mean
        if pd.api.types.is_numeric_dtype(credit[col]):
            credit[col] = credit[col].fillna(credit[col].mean())
        else:
            # Handle non-numeric columns (e.g., fill with a suitable value)
            # For example, fill with the most frequent value (mode):
            credit[col] = credit[col].fillna(credit[col].mode()[0])
            # or keep null if better
            #pass

credit.isnull().sum().sum()

np.int64(0)

In [ ]:
credit.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
# Distribution of Transactions
df=credit['Class'].value_counts(normalize=False)
print(df)

Class
0.00000    87034
1.00000      205
0.00235        1
Name: count, dtype: int64


This Dataset is highly unbalanced

*   0 represents Normal Transaction
*   1 represents Fraudent Transaction

In [ ]:
# Seperating data for analysis
legit=credit[credit.Class==00]
fraud=credit[credit.Class==1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(87034, 31)
(205, 31)


In [ ]:
# Statistical measure of the data
legit.Amount.describe()

count    87034.000000
mean        98.392549
std        267.475633
min          0.000000
25%          7.680000
50%         27.000000
75%         89.790000
max      19656.530000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     205.000000
mean      101.984439
std       228.446986
min         0.000000
25%         1.000000
50%         7.580000
75%        99.990000
max      1809.680000
Name: Amount, dtype: float64

In [ ]:
# Compare the values for both transaction.i.e the mean value of both the transactions
credit.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                     
0.00000  39219.680136 -0.249921 -0.050180  0.700945  0.151490 -0.267417   
0.00235  61596.000000 -6.759521  5.572257 -0.951531 -3.194834 -0.277593   
1.00000  34268.526829 -6.385940  4.394241 -8.432887  5.084061 -4.597691   

               V6        V7        V8        V9  ...       V20       V21  \
Class                                            ...                       
0.00000  0.099006 -0.094600  0.047273 -0.018269  ...  0.040984 -0.031313   
0.00235  0.094179 -0.110135  0.054041 -0.025376  ...  0.041802 -0.029472   
1.00000 -1.955100 -6.705734  2.927156 -3.042642  ...  0.389325  0.752293   

              V22       V23       V24       V25       V26       V27       V28  \
Class                                                                           
0.00000 -0.105935 -0.036580  0.009933  0.132438  0.026564 -0.000518  0.002005   
0.00235 -0.106031 -0.037054  0.009686  0.132652  0.026722  0.000801  0.002105   
1.00000 -0.146696 -0.238003 -0.095301  0.223491  0.093696  0.560829  0.044682   

             Amount  
Class                
0.00000   98.392549  
0.00235   98.400990  
1.00000  101.984439  

[3 rows x 30 columns]

Under Sampling

Build a sample data set containing similar distribution of normal transaction and the fraudulent Transaction


Number of Fraudulent Transaction-->205

In [ ]:
legit_sample=legit.sample(n=205)

Concatenating two DataFrame

In [ ]:
new_dataset=pd.concat([legit_sample,fraud],axis=0) # when axix is 0 the data set is added one by one. and all the 205 values will be added below the legit_sample
# and if axis =1 then the values will be added column wise

In [ ]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
5869    6392  1.156674  0.121389  0.833018  0.693529 -0.231582  0.116713   
47999  43454 -1.197392 -0.278791  2.333173 -2.054812 -1.392619 -0.391009   
53780  46142 -6.980796  5.218648 -1.831266 -0.899360 -2.117425 -1.371601   
51134  44811  1.317676  0.099902  0.125162  0.441350 -0.229091 -0.554948   
52010  45240 -0.790186 -3.477652  0.062873  1.693093 -2.107304  0.110501   

             V7        V8        V9  ...       V21       V22       V23  \
5869  -0.370913 -0.073656  1.457301  ... -0.204267 -0.298140  0.024234   
47999  1.142603 -0.133437  1.383901  ...  0.354528  0.904954  0.117971   
53780 -0.538679  1.391641  2.417020  ... -0.733452 -1.844138  0.403018   
51134 -0.075032 -0.052054  0.368783  ... -0.332676 -1.042961  0.053776   
52010  0.744965 -0.173867  0.970963  ...  0.403912 -0.904053 -0.841591   

            V24       V25       V26       V27       V28   Amount  Class  
5869  -0.436348  0.171166  0.227132 -0.013358  0.020642    38.99    0.0  
47999  0.365367  0.650423 -0.692274  0.103793 -0.064792   253.10    0.0  
53780  0.246650  0.782398 -0.181322 -0.566547  0.474761     7.75    0.0  
51134 -0.508914  0.269775  0.152467 -0.041689  0.006771     1.98    0.0  
52010  0.466621 -0.125599  0.267148 -0.184789  0.200657  1028.90    0.0  

[5 rows x 31 columns]

In [ ]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
83053  59669  0.326007  1.286638 -2.007181  2.419675 -1.532902 -1.432803   
83297  59777 -8.257111 -4.814461 -5.365307  1.204230 -3.347420 -1.331601   
83417  59840 -3.215382 -0.364223 -1.261883  3.794949  0.711206 -1.316476   
84543  60353 -3.975216  0.581573 -1.880372  4.319241 -3.024330  1.240793   
86155  61108 -2.756007  0.683821 -1.390169  1.501887 -1.165614 -0.131207   

             V7        V8        V9  ...       V21       V22       V23  \
83053 -2.459530  0.617738 -1.125861  ...  0.556895  0.169776 -0.174357   
83297 -1.967893  1.295438 -1.674415  ...  0.436390 -0.077553 -3.091624   
83417 -5.165141  0.625278 -1.582301  ...  0.401341  0.152191 -0.934675   
84543 -1.909559  0.660718 -2.752611  ...  0.578984  1.397311  1.045322   
86155 -1.478741 -0.246922 -0.100523  ...  0.320474  0.611027  0.174864   

            V24       V25       V26       V27       V28  Amount  Class  
83053  0.308061  0.710996 -0.231030  0.580495  0.300984    6.27    1.0  
83297 -0.390201 -0.288689 -0.340004  0.039819 -1.007900  319.20    1.0  
83417 -0.256148 -0.469403 -0.282496  0.866077 -0.433466    5.91    1.0  
84543 -0.304000  0.005295  0.235435  0.962015 -0.673557  454.82    1.0  
86155 -0.502151 -0.174713  1.179242 -1.166315  0.821215  101.50    1.0  

[5 rows x 31 columns]

In [ ]:
new_dataset['Class'].value_counts()

Class
0.0    205
1.0    205
Name: count, dtype: int64

In [ ]:
new_dataset.groupby('Class').mean()


Time        V1        V2        V3        V4        V5  \
Class                                                                   
0.0    36750.414634 -0.144384  0.051168  0.707965  0.224098 -0.410806   
1.0    34268.526829 -6.385940  4.394241 -8.432887  5.084061 -4.597691   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0.0    0.044859  0.001018 -0.001580  0.074324  ...  0.068110 -0.023873   
1.0   -1.955100 -6.705734  2.927156 -3.042642  ...  0.389325  0.752293   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0.0   -0.147112  0.012194 -0.048084  0.123181 -0.020599  0.015378  0.024137   
1.0   -0.146696 -0.238003 -0.095301  0.223491  0.093696  0.560829  0.044682   

           Amount  
Class              
0.0    107.193805  
1.0    101.984439  

[2 rows x 30 columns]

This setp is important to check wheather we got a good sample or bad sample adn if we got a bad sample so the mean values will be totally different

Here the nature of the dataset is not changed it means that difference is still there

Splitting  the data into Features and Targets

In [ ]:
X = new_dataset.drop(columns='Class',axis=1)
Y = new_dataset['Class']

In [ ]:
print(X)

        Time        V1        V2        V3        V4        V5        V6  \
5869    6392  1.156674  0.121389  0.833018  0.693529 -0.231582  0.116713   
47999  43454 -1.197392 -0.278791  2.333173 -2.054812 -1.392619 -0.391009   
53780  46142 -6.980796  5.218648 -1.831266 -0.899360 -2.117425 -1.371601   
51134  44811  1.317676  0.099902  0.125162  0.441350 -0.229091 -0.554948   
52010  45240 -0.790186 -3.477652  0.062873  1.693093 -2.107304  0.110501   
...      ...       ...       ...       ...       ...       ...       ...   
83053  59669  0.326007  1.286638 -2.007181  2.419675 -1.532902 -1.432803   
83297  59777 -8.257111 -4.814461 -5.365307  1.204230 -3.347420 -1.331601   
83417  59840 -3.215382 -0.364223 -1.261883  3.794949  0.711206 -1.316476   
84543  60353 -3.975216  0.581573 -1.880372  4.319241 -3.024330  1.240793   
86155  61108 -2.756007  0.683821 -1.390169  1.501887 -1.165614 -0.131207   

             V7        V8        V9  ...       V20       V21       V22  \
5869  -0.3709

In [ ]:
print(Y)

5869     0.0
47999    0.0
53780    0.0
51134    0.0
52010    0.0
        ... 
83053    1.0
83297    1.0
83417    1.0
84543    1.0
86155    1.0
Name: Class, Length: 410, dtype: float64


In [ ]:
# Split the data into training data and testing  data by using traing test split

In [ ]:
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.2,stratify=Y,random_state=2) # 0.2 represent 20% of test data and 80% is training data
# and if in stratify we will not assign Y so the distribution of 0 and 1 will be different in training data and testing data
# random_state means how we want to split our data

In [ ]:
print(X.shape,X_train.shape,X_test.shape)

(410, 30) (328, 30) (82, 30)


we have total 410 values and 328 are in training and 82 are in testing

Model Training

Logistic Regression- it is used for Binary Classification

In [ ]:
model=LogisticRegression()

In [ ]:
# Training the logistic regression modely with training data
model.fit(X_train,Y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

Model Evaluation


Accuracy Score

In [ ]:
# Accuracy on training  data
X_train_prediction=model.predict(X_train)
training_data_accuracy=accuracy_score(X_train_prediction,Y_train)

In [ ]:
print("Accuract on training data",training_data_accuracy)

Accuract on training data 0.967479674796748
